[Excel] Document API to load linked entity cell values (#5807)
OfficeDev/office-js-docs-pr: docs/excel/excel-data-types-linked-entity-cell-values.md

## Implement the linked entity load service function

> The add-in must provide a linked entity load service function to handle requests from Excel when property values are needed for any linked entity cell values. The function is identified with the `@linkedEntityLoadService` JSDoc tag. Depending on the version of the Excel client in which the add-in is installed, to load the linked entity cell values, implement one of the following options in your linked entity load service function.

Reviewer: Given this removal of the semi-annual channel, how much do we really want to emphasize older patterns?

Author: Good question. I'll follow up on this.

- Create helper functions to create the `LinkedEntityCellValue` objects.
- Register an event handler for the [onLinkedEntityCellValueLoaded](/javascript/api/excel/excel.linkedentitydatadomaincollection#excel-excel-linkedentitydatadomaincollection-onlinkedentitycellvalueloaded-member) event. Then, call [loadLinkedEntityCellValue](/javascript/api/excel/excel.linkedentitydatadomaincollection#excel-excel-linkedentitydatadomaincollection-loadlinkedentitycellvalue-member(1)), passing the linked entity ID as a parameter. The `loadLinkedEntityCellValue` API is supported in ExcelApi 1.21 and later.

The following code examples show how to create a function that uses each of these implementations to handle data requests from Excel for the **Products**, **Categories**, and **Suppliers** data domains. Select the tab for your preferred implementation option.

# [Helper function](#tab/helper-function)

### Create helper functions

> The following code sample shows the helper function to create a product linked entity cell value. This function is called by the `contosoLoadService` load service function to create a linked entity for a specific product ID. Notes on the following code.

Reviewer: To be consistent with the later phrasing.
```suggestion
The following code sample shows the helper function to create a product linked entity cell value. This function is called by the `contosoLoadService` load service function to create a linked entity for a specific product ID. Note the following about the code.
```

}
```

### Implement the load service function using the helper functions

The following code shows a linked entity load service function that calls the helper functions to create the linked entity cell values. Note the following about the code.

- The load service function parses the incoming `LinkedEntityLoadServiceRequest` object to extract the domain ID and entity IDs. These IDs are used to identify which data domain an entity belongs to, so that the appropriate helper functions can be called to create a linked entity cell value.
- Helper functions create the complete `LinkedEntityCellValue` objects with all properties populated.
- The load service function returns a `LinkedEntityLoadServiceResult` object containing the linked entity cell values in the same order they were requested.

```typescript
// Linked entity data domain constants
const productsDomainId = "products";
const categoriesDomainId = "categories";
const suppliersDomainId = "suppliers";

// Linked entity cell value constants
const addinDomainServiceId = 268436224;
const defaultCulture = "en-US";

/**

>  * Custom function which acts as the "service" or the data provider for a `LinkedEntityDataDomain`, that is

Reviewer: We're inconsistent with using code ticks in sample code in this article. Since this isn't getting parsed like our type definitions, I'd err in favor of not formatting code comments.

Author: Removed backticks from code comments.

>  * called on demand by Excel to resolve/refresh `LinkedEntityCellValue`s of that `LinkedEntityDataDomain`.

Reviewer: ```suggestion
 * called on demand by Excel to resolve and refresh `LinkedEntityCellValue`s of that `LinkedEntityDataDomain`.
```

 * @customfunction
 * @linkedEntityLoadService

>  * @param {any} request Request to resolve/refresh `LinkedEntityCellValue` objects.

Reviewer: ```suggestion
 * @param {any} request Request to resolve and refresh `LinkedEntityCellValue` objects.
```

>  * @return {any} Resolved/Refreshed `LinkedEntityCellValue` objects that were requested in the passed-in request.

Reviewer: ```suggestion
 * @return {any} Resolved or refreshed `LinkedEntityCellValue` objects that were requested in the passed-in request.
```

 */
function contosoLoadService(request: any): any {
    const notAvailableError = new CustomFunctions.Error(CustomFunctions.ErrorCode.notAvailable);
    console.log(`Fetching linked entities from request: ${request} ...`);

    try {
        // Parse the request that was passed-in by Excel.
        const parsedRequest: Excel.LinkedEntityLoadServiceRequest = JSON.parse(request);
        // Initialize result to populate and return to Excel.
        const result: Excel.LinkedEntityLoadServiceResult = { entities: [] };

        // Identify the domainId of the request and call the corresponding function to create
        // linked entity cell values for that linked entity data domain.
        for (const { entityId } of parsedRequest.entities) {

>             var linkedEntityResult = null;

Reviewer: ```suggestion
            let linkedEntityResult = null;
```

        // The resolver is a callback from the load service function that waits for the entity to load.
        const resolver = pendingEntityLoads.get(entityKey);
        if (resolver) {
            // Call the resolver with the loaded entity data.
            // This fulfills the Promise that the load service function was waiting on, allowing it to continue and add this entity to the results array.
            resolver(event.linkedEntityCellValue);
            // Clean up the map entry since we no longer need this resolver.
            pendingEntityLoads.delete(entityKey);
        }
    }
}
```

### Implement the load service function using the `loadLinkedEntityCellValue` API

Once you register and implement an event handler to identify when a linked entity cell value is loaded, call the `loadLinkedEntityCellValue` method for each entity in the load service function. Note the following about the code.

- The function must be tagged with `@linkedEntityLoadService`, so that Excel knows where to send the [LinkedEntityLoadServiceRequest](/javascript/api/excel/excel.linkedentityloadservicerequest) object.

> - Because the `loadLinkedEntityCellValue` method requires a `LinkedEntityId` as a parameter, you must first create a `LinkedEntityId` object for each entity. To create a `LinkedEntityId` object, use the entity's ID and the domain ID from the `LinkedEntityLoadServiceRequest` object.

Reviewer: Grammar suggestion to take or leave.
```suggestion
- The `loadLinkedEntityCellValue` method requires a `LinkedEntityId` as a parameter. You must first create a `LinkedEntityId` object for each entity. To create a `LinkedEntityId` object, use the entity's ID and the domain ID from the `LinkedEntityLoadServiceRequest` object.
```

- Once the linked entity cell value finishes loading, the `onLinkedEntityCellValueLoaded` event occurs and the `handleLinkedEntityLoaded` handler runs to get the linked entity cell value.
- The load service function returns a [LinkedEntityLoadServiceResult](/javascript/api/excel/excel.linkedentityloadserviceresult) object back to Excel to resolve or refresh the linked entity cell values. Note that the load service function must return the loaded linked entity cell values in the same order as their entities in the `LinkedEntityLoadServiceRequest` object.

```typescript
// Linked entity data domain constants
const productsDomainId = "products";
const categoriesDomainId = "categories";
const suppliersDomainId = "suppliers";

// Linked entity cell value constants
const addinDomainServiceId = 268436224;

// Reuse pendingEntityLoads declared in the earlier event-handler snippet.

/**
 * Custom function which acts as the "service" or the data provider for a `LinkedEntityDataDomain`, that is

>  * called on demand by Excel to resolve/refresh `LinkedEntityCellValue`s of that `LinkedEntityDataDomain`.

Reviewer: ```suggestion
 * called on demand by Excel to resolve and refresh `LinkedEntityCellValue`s of that `LinkedEntityDataDomain`.
```

 * @customfunction
 * @linkedEntityLoadService

>  * @param {any} request Request to resolve/refresh `LinkedEntityCellValue` objects.

Reviewer: ```suggestion
 * @param {any} request Request to resolve and refresh `LinkedEntityCellValue` objects.
```

>  * @return {any} Resolved/Refreshed `LinkedEntityCellValue` objects that were requested in the passed-in request.

Reviewer: ```suggestion
 * @return {any} Resolved or refreshed `LinkedEntityCellValue` objects that were requested in the passed-in request.
```

}
```

> ## Query linked entity cell values

Author: @AlexJerabek Moved the load API to its own section after receiving feedback from the feature crew. Please take another look when you get a chance and let me know if you have any additional feedback. Thanks!
